API.md

```diff
@@ -90,3 +90,65 @@ _Note: The `message` field is meant only to describe the error to a user, this t
 
 `{"error":{"id":"no_login","message":"You are not logged in to an account"},"status":"error"}`
 
+### `/resetpass` - Reset a password for a user
+#### Params
+- `account`
+- `email`
+
+#### Responses
+##### Success
+
+`{"status":"ok"}`
+
+#### Errors
+
+`{"error":{"id":"no_account","message":"Unable to find matching account"},"status":"error"}`
```

Comment:
Password reset tools should not expose either that an account exists or that the email or otherwise was (in)correct. Doing so exposes an attack vector for finding valid accounts.

Reply:
Ah I'm assuming you mean I should collapse no_account and wrong_code to something like reset_failed ?

Reply:
In this part of the API (the `/resetpass` endpoint), `no_account` means that something didn't match, either the email or the username, doesn't say which

Reply:
Still informs the requester that the data does not exist, which therefore allows a search of what _does_ exist

Reply:
Should I just "allow" it through and return status:ok for invalid data? I know some sites do that for stuff

Reply:
Yeah thats what I was thinking, a "if this was correct you should get an email with further instructions"

Reply:
This is an internal API, it should absolutely not mask information. It is up to how we present the data to the user to mask the information - which this message is not used for.

@A-UNDERSCORE-D I understand that you want this to be air tight in 100% of all use cases and that people on our own team should not be trusted. But this is an extreme viewpoint that only causes issues when it comes to debugging issues further down the line internally. We must be pragmatic about these things.

Reply:
@prawnsalad All APIs regardless of location should not trust their clients. What happens if this is exposed to the internet at some point? then what? "oops sorry we provided userdata to the internet"? Debugging should not occur over an API. These things log locally for a reason. There's a reason, for example, HTTP servers (including only internally accessible ones) generally return `500 Internal Server Error` and not `500 Line x in file y broke because z`. In general security through obscurity (ie, "its an internal API") is a very dangerous thing to rely on. 

>I understand that you want this to be air tight in 100% of all use cases and that people on our own team should not be trusted
>We must be pragmatic about these things.

So, in other words "security is second to my ability to debug production systems in insecure ways"?

I see no reason to not trust your team. That's fine. However the things using this API are not your team, they're code. Code is not infallible, in fact, code in general almost always has bugs.

Bottom line: If an API fails to behave as expected, it should be tested in a clean environment. If your concern is debugging, perhaps it should be unit tested so bugs don't crop up, and when and if bugs do crop up, a test network is more than sufficient to provide testing. A common practice in network security is to write code et al. as if you are already compromised, and want to mitigate damage. This is also known as [defensive development](https://en.wikipedia.org/wiki/Defensive_programming). These practices should be applied to anything that requires high security (ie, works in any way shape or form with userdata), lest the GDPR come to bite us. 

This is not me trying to be difficult, this is not me being extreme. Me being extreme would be me pushing that every single server be connected over a VPN and this API (with EVERY security consideration I've already mentioned in place) only accessible via that VPN AND through firewalls on the specific machine. Security of thousands of users is not something to be taken lightly. While that would be nice, it would also be very difficult to bring existing networks into.

Reply:
Resolved by 186bc64e99c05166c6cee5361ba8539d992108cc
@prawnsalad For debugging, the specific error is logged internally and can be checked there.

Reply:
re. VPNs and firewalls, this is something that I have already put onto the todo list since nobody in previous years seems to have considered that enough to put it in place.

re. debugging, this isn't about putting debugging into API. This is about putting relevant debugging information into the correct places as not to waste peoples time. If people start complaining that they cannot login, then if irc.com devs cannot see generic errors, they then have to go a level deeper and ask network level staff for more logs. You may not care about this, but we do as this is when it starts costing money, time, and can even delay debugging if everybody is not around. This is not just our use case - this is a very common thing to bare in mind for any project.

I fully understand the angle you are coming from, but you must also understand that a volunteer non-commercial project has entirely different expectations from a commercial project. You may not care so much for debugging a production system when random issues happen that may not appear during a staging test, but I and future organisations using the network do.

Reply:
>then if irc.com devs cannot see generic errors, they then have to go a level deeper and ask network level staff for more logs

Why is this an issue? especially when, if you provide errors like "this password doesn't match this account, its a very nice target for bruteforcing. If the API is exposed globally, even more so. The data is there for IRC.com devs. You want us to work together, you keep saying so, so why cant we?

Also, if me, as a volunteer for a "volunteer non-commercial project" has tighter security requirements than you, a CEO of a commercial project which will handle "many more users", then I think you have a problem.

Reply:
also:
>You may not care so much for debugging a production system when random issues happen that may not appear during a staging test, but I and future organisations using the network do.

I do care. There are provisions for debugging. They're just not exposed over the API.

Reply:
>This is not just our use case - this is a very common thing to bare in mind for any project.

Not in any project that I've seen. I'm pushing for industry standard security practices.

Reply:
> Why is this an issue? especially when, if you provide errors like "this password doesn't match this account", its a very nice target for bruteforcing. If the API is exposed globally, even more so.

Again, this is an internal API. We won't be brute forcing our own users. This API is never going to be made public as it's an internal API. I'm not sure how this can be made any more clear.

> The data is there for IRC.com devs. You want us to work together, you keep saying so, so why cant we?

I'm not sure what you're saying... the relevant snoonet staff have access to the same data if not more.

> Also, if me, as a volunteer for a "volunteer non-commercial project" has tighter security requirements than you, a CEO of a commercial project which will handle "many more users", then I think you have a problem.

You are not understanding the differences between the security of an internal system VS an external public accessible system. I wouldn't go around suggesting people have problems when you don't have the full knowledge of both sides of the argument. There are individual teams, web and network. Web teams don't generally have access to network related logs. However, if logs are failing then it's the web team that would first get the complaints. If the web team can see via web service logs that it's an incorrect password failure, they can react on that. if they see that it's a code failure issue, they can react on that. If it's on any other issue, they can react on that. Depending on the type of error, they may invoke the network team to look further. If the error is too vague and they depend on the network team to dig into logs but the network team is not available at that time, this could be many hours wasted if it's a simple error that could have been handled already.

This is how working with multiple internal teams work. Limiting the data - sure. Proving just enough data to help debugging between teams goes a long way.

***Again - this is all internal data.*** irc.com *and snoonet staff* already know if an account exists or not. Providing this information for debugging is not leaking anything.

Reply:
Also, lets be honest. This API, internal or not, is more secure and privacy focussed than what's publicly available to anybody out there.

> NickServ(NickServ@services)- Nick wedwefgrtyhj isn't registered.

Reply:
>Again, this is an internal API. We won't be brute forcing our own users. This API is never going to be made public as it's an internal API. I'm not sure how this can be made any more clear.

It exists on the internet. It can be attacked. Yes its firewalled and authenticated. That. Doesnt. Matter. The auth code could have a bug in it. The python frontend could be broken. Your backend could be broken. the API keys could be stolen. This happens. Ask any number of compromised software

In fact, the fact that it exists on the internet, and is *designed* to be used as an external API, means that it is **not** an internal API. It simply currently is only accessed by internal people and code. (which again, can be compromised)

_**Act as if you are already compromised and working to mitigate an attack**_
This doesnt just apply to external code, in fact, it can NOT just apply external code. It must apply to the entire infrastructure.

>I'm not sure what you're saying... the relevant snoonet staff have access to the same data if not more.

You said, and I quote:
>...then if irc.com devs cannot see generic errors, they then have to go a level deeper and ask network level staff for more logs.

Hence the question regarding working together.

>You are not understanding the differences between the security of an internal system VS an external public accessible system.

Internal systems should be as secure if not more secure. Also. **It is on the internet** and even IF it was on an internal network that was 100% airgapped. Malicious code and parties exist. 

>Web teams don't generally have access to network related logs. However, if logs are failing then it's the web team that would first get the complaints. If the web team can see via web service logs that it's an incorrect password failure, they can react on that. if they see that it's a code failure issue, they can react on that. If it's on any other issue, they can react on that. Depending on the type of error, they may invoke the network team to look further. If the error is too vague and they depend on the network team to dig into logs but the network team is not available at that time, this could be many hours wasted if it's a simple error that could have been handled already.

A extra time is worth not exposing what could possibly release user data.

>Also, lets be honest. This API, internal or not, is more secure and privacy focussed than what's publicly available to anybody out there.
>>NickServ(NickServ@services)- Nick wedwefgrtyhj isn't registered.

IRC already has existing logging, flood protections, and monitoring. I don't see how this is relevant. Not to mention that `/ns info`-ing a nick isn't the same as bruteforcing password reset tokens, or passwords in general

Not to mention the fact, that NickServ itself will KILL users off after attacks like this API provides the ability to do. And the network will kill users off when flooding (see aforementioned flood protection)

# bottom line

This supposedly "internal" API, as demonstrated above, is nether internal nor intended to be internal. It just happens to be currently used by internal people and code (which, again, again, can be compromised)

And, if, and I really hope it never happens, but IF this is compromised, and leads to an investigation and fine due to GDPR and other laws regarding user privacy, do we really want to think back to this and say "perhaps we should have implemented more security"?

Reply:
I've read, understood, and noted your comments.

Of course, if any fine or investigation is made towards Snoonet or irc.com, I am the one ultimately responsible. I've mentioned this before. I fully realise that I am also taking on the responsibility of all the previous fuckups from Snoonet staff. The good or bad.

Based on previous conversations I don't expect you nor I to see eye to eye on these topics for the most part, but we should at least respect each others input and role in the network. And one of the important factors at this stage in this whole platform is the ease of debugging and development as we push forward, which is where these error messages come into play.

Reply:
# bottom line x2
oops, missed this attention grabbing title in my previous comment

Reply:
Ok, if you have a legitimate change to request, request changes on the diff. Otherwise I'm going to lock the thread.